server/user/user.controller.js

```diff
@@ -1,13 +1,25 @@
 'use strict';
 
 const { createError, validationError } = require('../shared/error/helpers');
-const { NOT_FOUND } = require('http-status-codes');
+const { NO_CONTENT, NOT_FOUND } = require('http-status-codes');
+const { role } = require('../../config/shared');
 const { User } = require('../shared/database');
 
-function index(req, res) {
-  const attributes = ['id', 'email', 'role'];
-  return User.findAll({ attributes })
-    .then(users => res.json({ data: users }));
+function index({ query: { roleType } }, res) {
+  let options = { attributes: ['id', 'email', 'role'], raw: true };
+  if (roleType) options.where = { role: { $in: role.getRoleNames(roleType) } };
+  return User.findAll(options).then(data => res.json({ data }));
+}
+
+function upsert({ body: { email, role } }, res) {
+  return User.findOne({ where: { email }, paranoid: false })
+    .then(user => user ? user.update({ role }) : User.invite({ email, role }))
+    .then(user => user.deletedAt ? user.restore() : user)
```

Comment:
What bothers me here is that if you do an update of soft-deleted user you'll end up with 3 queries:
1. `SELECT` to find it
2. `UPDATE` to patch it
3. `UPDATE` to restore it

Can we merge last two? :thinking: